packages/angular_devkit/schematics/src/engine/task.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { PriorityQueue } from '@angular-devkit/core';
+import { Observable } from 'rxjs/Observable';
+import { SchematicContext } from './interface';
+
+
+export interface TaskConfiguration<T = {}> {
+  name: string;
+  dependencies?: Array<TaskId>;
+  options?: T;
+}
+
+export interface TaskConfigurationGenerator<T = {}> {
+  toConfiguration(): TaskConfiguration<T>;
+}
+
+export type TaskExecutor<T = {}>
+  = (options: T | undefined, context: SchematicContext) => Observable<void>;
+
+export interface TaskExecutorFactory<T> {
+  readonly name: string;
+  create(options?: T): Promise<TaskExecutor>;
+}
+
+export interface TaskId {
```

Comment:
Seems like the engine should keep a `Map<TaskId, TaskInfo>` cache and `TaskId` should just be the number. Otherwise everything needs to be readonly, which in this case it isn't.

Reply:
That was my original plan but was concerned about too many types.  But it does provide much better encapsulation.

Done.